common/src/collision.rs

```diff
@@ -0,0 +1,477 @@
+
+                Self::add_sub_bb(
+                    &mut bounding_boxes,
+                    ChunkBoundingBox::get_chunk_bounding_box(
+                        node.unwrap(),
+                        Vertex::from_sides(sides[0], sides[1], sides[2]).unwrap(), // pretty sure this is pointless
+                        translated_position,
+                        radius,
+                        dimension,
+                    ),
+                );
+            }
+
+            // get BB for chunk sharing only a vertex.
+            Self::add_sub_bb(
+                &mut bounding_boxes,
+                ChunkBoundingBox::get_chunk_bounding_box(
+                    opposite_node,
+                    Vertex::from_sides(sides[0], sides[1], sides[2]).unwrap(),
+                    opposite_position,
+                    radius,
+                    dimension,
+                ),
+            );
+        }
+
+        BoundingBox { bounding_boxes }
+    }
+
+    // adds a sub-bounding box to a list if it exists
+    pub fn add_sub_bb(list: &mut Vec<ChunkBoundingBox>, result: Option<ChunkBoundingBox>) {
+        if let Some(x) = result {
+            list.push(x);
+        }
+    }
+
+    pub fn every_voxel_address<'a>(&'a self) -> impl Iterator<Item = VoxelAddress> + 'a {
+        self.bounding_boxes.iter().flat_map(|cbb| {
+            cbb.every_voxel().map(move |index| VoxelAddress {
+                node: cbb.node,
+                chunk: cbb.chunk,
+                index,
+            })
+        })
+    }
+
+    pub fn display_summary(&self) {
+        let number_of_chunk_bouding_boxes = {
+            let mut n = 0_i32;
+            for _cbb in self.bounding_boxes.iter() {
+                n += 1;
+            }
+            n
+        };
+        println!(
+            "Bounding box spanning {} chunks:",
+            number_of_chunk_bouding_boxes
+        );
+        for cbb in self.bounding_boxes.iter() {
+            println!("\tA chunk");
+            //println!("\t\twith node id {}", cbb.node); // can't easily display node id
+            println!(
+                "\t\twith bounding box stretching from ({}, {}, {}) to ({}, {}, {})",
+                cbb.min_xyz[0],
+                cbb.min_xyz[1],
+                cbb.min_xyz[2],
+                cbb.max_xyz[0],
+                cbb.max_xyz[1],
+                cbb.max_xyz[2]
+            );
+        }
+    }
+}
+
+// translated_position should be the object position in the node coordinates of the chunk.
+// node can be easily factored out if it stops being convienent.
+impl ChunkBoundingBox {
+    pub fn get_chunk_bounding_box(
+        node: NodeId,
+        chunk: Vertex,
+        translated_position: na::Vector4<f64>,
+        radius: f64,
+        dimension: u8,
+    ) -> Option<Self> {
+        let euclidean_position =
+            (chunk.chunk_to_node().try_inverse().unwrap() * translated_position).xyz();
```

Comment:
The `chunk_to_node` method works with homogeneous coordinates, and due to the transition between hyperbolic and Euclidean geometry, the "w" coordinate in general will not be 1. Therefore, dividing by the "w" coordinate is necessary here. I believe this is one of the two main sources of error for the collision checking.

Sample code:
```rust
let euclidean_position = chunk.chunk_to_node().try_inverse().unwrap() * translated_position;
let euclidean_position = euclidean_position.xyz() / euclidean_position[3];
```